emmarin/cl/cl/src/cl/indexed.rs

```diff
@@ -0,0 +1,325 @@
+        }
+    }
+
+    fn hashed_leaves(&self) -> Vec<[u8; 32]> {
+        merkle::padded_leaves(&self.leaves.iter().map(|l| l.to_bytes()).collect::<Vec<_>>())
+    }
+
+    pub fn root(&self) -> [u8; 32] {
+        merkle::root(&self.hashed_leaves())
+    }
+
+    pub fn insert(&mut self, value: Nullifier) -> UpdateProof {
+        let (idx, &low_nf) = self
+            .leaves
+            .iter()
+            .enumerate()
+            .find(|(_, l)| in_interval(**l, value))
+            .expect("element already exist");
+        let low_nf_path = merkle::path(&self.hashed_leaves(), idx);
+
+        let new_leaf = Leaf {
+            value,
+            next_value: low_nf.next_value,
+        };
+        self.leaves[idx].next_value = value;
+
+        let mut mmr = MMR::new();
+        for leaf in &self.leaves {
+            mmr.push(&leaf.to_bytes());
+        }
+        assert_eq!(self.root(), frontier_root(&mmr.roots));
+
+        self.leaves.push(new_leaf);
+
+        UpdateProof {
+            value,
+            low_nf_path,
+            low_nf,
+            mmr,
+        }
+    }
+
+    pub fn insert_batch(&mut self, mut values: Vec<Nullifier>) -> BatchUpdateProof {
+        values.sort();
+
+        let mut low_nfs_idx = <BTreeMap<_, Vec<_>>>::new();
+
+        for value in &values {
+            let idx = self
+                .leaves
+                .iter()
+                .enumerate()
+                .find(|(_, l)| in_interval(**l, *value))
+                .expect("element already exist")
+                .0;
+            low_nfs_idx.entry(idx).or_default().push(*value);
+        }
+
+        let mut new_leaves = Vec::new();
+        let mut low_nf_paths = Vec::new();
+        let mut low_nfs = Vec::new();
+        for (idx, values) in &low_nfs_idx {
+            let low_nf = self.leaves[*idx];
+            low_nfs.push(low_nf);
+            low_nf_paths.push(merkle::path(&self.hashed_leaves(), *idx));
+            self.leaves[*idx].next_value = *values.first().unwrap();
+            for w in values.windows(2) {
+                let prev = w[0];
+                let next = w[1];
+                new_leaves.push(Leaf {
+                    value: prev,
+                    next_value: next,
+                });
+            }
+            new_leaves.push(Leaf {
+                value: *values.last().unwrap(),
+                next_value: low_nf.next_value,
+            });
+        }
+
+        let mut mmr = MMR::new();
+        for leaf in &self.leaves {
+            mmr.push(&leaf.to_bytes());
+        }
+        assert_eq!(self.root(), frontier_root(&mmr.roots));
+
+        for new_leaf in new_leaves {
+            self.leaves.push(new_leaf);
+        }
+
+        BatchUpdateProof {
+            values,
+            low_nfs,
+            low_nf_paths,
+            mmr,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Copy, Serialize, Deserialize)]
+pub struct Leaf {
+    value: Nullifier,
+    next_value: Nullifier,
+}
```

Comment:
we don't need the next_index here? 🤔

Reply:
It's implicit in the data what is the next leaf index and it's not needed by the proof anyway. Executor-side implementation can use some helper ds to find the low_nf faster than a full scan